File: apk/app/src/main/java/com/example/momentra/ui/shell/empty/group/GroupBudgetUtils.kt
package com.example.momentra.ui.shell.empty.group

import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.util.Locale

object GroupBudgetUtils {
    const val CUSTOM_OPTION = "Custom…"
    val PRESET_OPTIONS = listOf("₹80,000", "₹50,000", "₹1,20,000", "₹25,000", CUSTOM_OPTION)

    /** Parses display strings like `₹1,20,000` into API decimal `120000`. */
    fun parseDisplayToApiAmount(display: String): String? {
        val digits = display
            .replace("₹", "")
            .replace(",", "")
            .replace(" ", "")
            .trim()
        if (digits.isEmpty()) return null
        if (!digits.matches(Regex("^\\d+(\\.\\d{1,4})?$"))) return null
        return digits
    }

    fun resolveBudgetAmount(displayBudget: String, customAmount: String): String? {
        if (displayBudget == CUSTOM_OPTION) {
            return parseDisplayToApiAmount(customAmount)
        }
        return parseDisplayToApiAmount(displayBudget)
    }

    fun formatApiAmountForDisplay(amount: String, currencyCode: String = "INR"): String {
        val value = amount.toDoubleOrNull() ?: return amount
        val prefix = when (currencyCode) {
            "INR" -> "₹"
            "USD" -> "$"
            "EUR" -> "€"
            else -> "$currencyCode "
        }
        val symbols = DecimalFormatSymbols(Locale.US).apply { groupingSeparator = ',' }
        val pattern = if (value % 1.0 == 0.0) "#,##0" else "#,##0.00"
        return prefix + DecimalFormat(pattern, symbols).format(value)
    }

    fun formatCustomAmountInput(raw: String): String {
        val digits = raw.filter { it.isDigit() }
        if (digits.isEmpty()) return ""
        val num = digits.toLongOrNull() ?: return ""
        val symbols = DecimalFormatSymbols(Locale.US).apply { groupingSeparator = ',' }
        return DecimalFormat("#,###", symbols).format(num)
    }

    fun summaryLabel(displayBudget: String, customAmount: String): String {
        if (displayBudget == CUSTOM_OPTION) {
            val parsed = parseDisplayToApiAmount(customAmount)
            return if (parsed != null) formatApiAmountForDisplay(parsed) else "Custom amount"
        }
        return displayBudget
    }
}

File: momentra/momentra/Shell/GroupEmpty/GroupBudgetUtils.swift
import Foundation

enum GroupBudgetUtils {
    static let customOption = "Custom…"
    static let presetOptions = ["₹80,000", "₹50,000", "₹1,20,000", "₹25,000", customOption]
    static let purchaseAmountOptions = ["₹25,000", "₹50,000", "₹1,00,000", customOption]
    static let livingBudgetOptions = ["₹25,000", "₹40,000", "₹60,000", customOption]

    static func parseDisplayToApiAmount(_ display: String) -> String? {
        let digits = display
            .replacingOccurrences(of: "₹", with: "")
            .replacingOccurrences(of: ",", with: "")
            .replacingOccurrences(of: " ", with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !digits.isEmpty else { return nil }
        let pattern = "^\\d+(\\.\\d{1,4})?$"
        guard digits.range(of: pattern, options: .regularExpression) != nil else { return nil }
        return digits
    }

    static func resolveBudgetAmount(displayBudget: String, customAmount: String) -> String? {
        if displayBudget == customOption {
            return parseDisplayToApiAmount(customAmount)
        }
        return parseDisplayToApiAmount(displayBudget)
    }

    static func formatApiAmountForDisplay(_ amount: String, currencyCode: String = "INR") -> String {
        guard let value = Double(amount) else { return amount }
        let prefix: String = switch currencyCode {
        case "INR": "₹"
        case "USD": "$"
        case "EUR": "€"
        default: "\(currencyCode) "
        }
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.groupingSeparator = ","
        formatter.maximumFractionDigits = value.truncatingRemainder(dividingBy: 1) == 0 ? 0 : 2
        let body = formatter.string(from: NSNumber(value: value)) ?? amount
        return "\(prefix)\(body)"
    }

    static func formatCustomAmountInput(_ raw: String) -> String {
        let digits = raw.filter(\.isNumber)
        guard !digits.isEmpty, let num = Int64(digits) else { return "" }
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.groupingSeparator = ","
        formatter.maximumFractionDigits = 0
        return formatter.string(from: NSNumber(value: num)) ?? digits
    }

    static func summaryLabel(displayBudget: String, customAmount: String) -> String {
        if displayBudget == customOption {
            if let parsed = parseDisplayToApiAmount(customAmount) {
                return formatApiAmountForDisplay(parsed)
            }
            return "Custom amount"
        }
        return displayBudget
    }
}
